File: js/src/core/nms.ts
export function classAwareNMS(
  boxes: Float32Array, // flat (N*4) xyxy
  scores: Float32Array, // (N)
  classes: Int32Array, // (N)
  iouThreshold = 0.45,
): Int32Array {
  const N = scores.length;
  if (N === 0) return new Int32Array(0);

  // Group by class
  const byClass = new Map<number, number[]>();
  for (let i = 0; i < N; i++) {
    const c = classes[i] as number;
    let arr = byClass.get(c);
    if (!arr) {
      arr = [];
      byClass.set(c, arr);
    }
    arr.push(i);
  }

  const keepGlobal: number[] = [];
  for (const [, idxs] of byClass) {
    const kept = nmsSingleClass(boxes, scores, idxs, iouThreshold);
    keepGlobal.push(...kept);
  }
  // Sort by descending global score
  keepGlobal.sort((a, b) => (scores[b] as number) - (scores[a] as number));
  return new Int32Array(keepGlobal);
}

function nmsSingleClass(
  boxes: Float32Array,
  scores: Float32Array,
  idxs: number[],
  iou: number,
): number[] {
  // Sort idxs by score desc
  idxs.sort((a, b) => (scores[b] as number) - (scores[a] as number));
  const keep: number[] = [];
  const suppressed = new Set<number>();
  for (const i of idxs) {
    if (suppressed.has(i)) continue;
    keep.push(i);
    const ix1 = boxes[i * 4] as number;
    const iy1 = boxes[i * 4 + 1] as number;
    const ix2 = boxes[i * 4 + 2] as number;
    const iy2 = boxes[i * 4 + 3] as number;
    const iArea = Math.max(0, ix2 - ix1) * Math.max(0, iy2 - iy1);
    for (const j of idxs) {
      if (j === i || suppressed.has(j)) continue;
      const jx1 = boxes[j * 4] as number;
      const jy1 = boxes[j * 4 + 1] as number;
      const jx2 = boxes[j * 4 + 2] as number;
      const jy2 = boxes[j * 4 + 3] as number;
      const jArea = Math.max(0, jx2 - jx1) * Math.max(0, jy2 - jy1);
      const xx1 = Math.max(ix1, jx1);
      const yy1 = Math.max(iy1, jy1);
      const xx2 = Math.min(ix2, jx2);
      const yy2 = Math.min(iy2, jy2);
      const w = Math.max(0, xx2 - xx1);
      const h = Math.max(0, yy2 - yy1);
      const inter = w * h;
      const u = iArea + jArea - inter;
      if (u > 0 && inter / u > iou) suppressed.add(j);
    }
  }
  return keep;
}

File: python/src/yolo26_kit/core/nms.py
"""Class-aware NMS for non-e2e YOLO output. Pure numpy, no deps."""
from __future__ import annotations

from typing import Any

import numpy as np
from numpy.typing import NDArray


def class_aware_nms(
    boxes: NDArray[Any],     # (N, 4) xyxy
    scores: NDArray[Any],    # (N,)
    classes: NDArray[Any],   # (N,)
    iou_threshold: float = 0.45,
) -> NDArray[np.int64]:
    """Class-aware NMS. Returns indices to keep, sorted by descending score."""
    if boxes.size == 0:
        return np.empty((0,), dtype=np.int64)
    keep_all: list[int] = []
    for cls in np.unique(classes):
        mask = classes == cls
        idxs = np.nonzero(mask)[0]
        cls_boxes = boxes[mask]
        cls_scores = scores[mask]
        keep_in_cls = _nms_single_class(cls_boxes, cls_scores, iou_threshold)
        keep_all.extend(idxs[keep_in_cls].tolist())
    keep = np.asarray(keep_all, dtype=np.int64)
    # Re-sort by descending global score
    order = np.argsort(-scores[keep])
    return keep[order]


def _nms_single_class(
    boxes: NDArray[Any], scores: NDArray[Any], iou_threshold: float
) -> NDArray[np.int64]:
    if boxes.shape[0] == 0:
        return np.empty((0,), dtype=np.int64)
    x1 = boxes[:, 0]
    y1 = boxes[:, 1]
    x2 = boxes[:, 2]
    y2 = boxes[:, 3]
    areas = np.maximum(0.0, x2 - x1) * np.maximum(0.0, y2 - y1)
    order = np.argsort(-scores)
    keep: list[int] = []
    while order.size > 0:
        i = int(order[0])
        keep.append(i)
        if order.size == 1:
            break
        xx1 = np.maximum(x1[i], x1[order[1:]])
        yy1 = np.maximum(y1[i], y1[order[1:]])
        xx2 = np.minimum(x2[i], x2[order[1:]])
        yy2 = np.minimum(y2[i], y2[order[1:]])
        w = np.maximum(0.0, xx2 - xx1)
        h = np.maximum(0.0, yy2 - yy1)
        inter = w * h
        ious = inter / (areas[i] + areas[order[1:]] - inter + 1e-12)
        order = order[1:][ious <= iou_threshold]
    return np.asarray(keep, dtype=np.int64)
